File: 235-lowest-common-ancestor-of-a-binary-search-tree/235-lowest-common-ancestor-of-a-binary-search-tree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

struct TreeNode* lowestCommonAncestor(struct TreeNode* root, struct TreeNode* p, struct TreeNode* q) {
    
    if(root == NULL)
        return root;
    
    struct TreeNode* res;
    
    if(root->val < p->val && root->val < q->val)
        res = lowestCommonAncestor(root->right, p, q);
        
    else if (root->val > p->val && root->val > q->val)
        res = lowestCommonAncestor(root->left, p, q);
    else
        return res = root;
    
    return res;   
}

File: 235-lowest-common-ancestor-of-a-binary-search-tree/235-lowest-common-ancestor-of-a-binary-search-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val   int
 *     Left  *TreeNode
 *     Right *TreeNode
 * }
 */

func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
    if root == nil {
        return nil
    }
    
    var res *TreeNode
    
    if root.Val < p.Val && root.Val < q.Val {
        res = lowestCommonAncestor(root.Right, p, q)
    } else if root.Val > p.Val && root.Val > q.Val {
        res = lowestCommonAncestor(root.Left, p, q)
    } else {
        res = root
    }
    return res
}